File: frontend/js/views/years_to_fi.js
// Years to FI — savings rate, FI number (expenses / SWR), and the years for
// current savings + the annual surplus to reach it, via /calc/years-to-fi.
// Updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const FIELDS = [
    ['current_savings_usd', 'Current savings ($)', 50000],
    ['annual_income_usd', 'Annual income ($)', 80000],
    ['annual_expenses_usd', 'Annual expenses ($)', 40000],
    ['annual_return_pct', 'Expected real return (%)', 5],
    ['safe_withdrawal_rate_pct', 'Safe withdrawal rate (%)', 4],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });
const VIEW = 'years-to-fi';
let lastReport = null;
let lastBody = null;

export async function renderYearsToFi(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.ytf.h1.title">// YEARS TO FINANCIAL INDEPENDENCE</span></h1>
        <p class="muted small" data-i18n="view.ytf.hint.intro">
            FI is when your portfolio funds your spending indefinitely at a safe withdrawal rate
            — the FI number = annual expenses ÷ SWR (4% → 25× expenses). Starting from current
            savings and adding the income-minus-expenses gap each year (grown at your return),
            this finds the years to reach it. The dominant lever is the savings rate, far more
            than the return. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.ytf.h2.inputs">Your finances</h2>
            <form id="ytf-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.ytf.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
            <div id="ytf-tools" class="ce-toolbar"></div>
            <button type="button" id="ytf-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="ytf-sens" class="ce-sens"></div>
        </div>
        <div id="ytf-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#ytf-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        return body;
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcYearsToFi(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.ytf.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#ytf-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'years-to-fi.csv' });
    mount.querySelector('#ytf-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['years_to_fi', r.already_fi ? 0 : (r.years_to_fi == null ? '' : r.years_to_fi)],
        ['savings_rate_pct', r.savings_rate_pct],
        ['fi_number_usd', r.fi_number_usd],
        ['annual_savings_usd', r.annual_savings_usd],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#ytf-result');
    // Line chart: years to FI as annual expenses sweep 0.5x -> 1.5x (lower expenses both
    // raise the savings rate and shrink the FI number, so years fall steeply).
    const base = body.annual_expenses_usd || 40000;
    const xs = enh.linspace(base * 0.5, base * 1.5, 13);
    const pts = await Promise.all(xs.map(async (e) => {
        const rr = await api.calcYearsToFi({ ...body, annual_expenses_usd: e });
        return { x: e / 1000, y: rr && !rr.already_fi && rr.years_to_fi != null ? rr.years_to_fi : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'expenses $k', ylabel: 'years' });
    let yearsVal, yearsCls;
    if (r.already_fi) {
        yearsVal = t('view.ytf.already');
        yearsCls = 'pos';
    } else if (r.years_to_fi == null) {
        yearsVal = t('view.ytf.never');
        yearsCls = 'neg';
    } else {
        yearsVal = `${r.years_to_fi} ${t('view.ytf.years')}`;
        yearsCls = r.years_to_fi <= 15 ? 'pos' : r.years_to_fi <= 30 ? '' : 'neg';
    }
    const srCls = r.savings_rate_pct >= 50 ? 'pos' : r.savings_rate_pct >= 20 ? '' : 'neg';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.ytf.h2.result">The timeline</h2>
            <div class="cards">
                <div class="card ${yearsCls}"><div class="label" data-i18n="view.ytf.card.years">Years to FI</div>
                    <div class="value ${yearsCls}">${yearsVal}</div></div>
                <div class="card ${srCls}"><div class="label" data-i18n="view.ytf.card.rate">Savings rate</div>
                    <div class="value ${srCls}">${Number(r.savings_rate_pct).toFixed(1)}%</div></div>
                <div class="card"><div class="label" data-i18n="view.ytf.card.finumber">FI number</div>
                    <div class="value">${money(r.fi_number_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.ytf.card.savings">Annual savings</div>
                    <div class="value">${money(r.annual_savings_usd)}</div></div>
            </div>
            ${chart}
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#ytf-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: annual expenses 0.5x -> 1.5x; y: real return 1% -> 10%. Output: years to FI (lower better -> negate).
    const ex = base.annual_expenses_usd || 40000;
    const xVals = enh.linspace(ex * 0.5, ex * 1.5, 5);
    const yVals = enh.linspace(1, 10, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'annual_expenses_usd', yKey: 'annual_return_pct', xVals, yVals, compute: (b) => api.calcYearsToFi(b), pick: (r) => (r && !r.already_fi && r.years_to_fi != null ? r.years_to_fi : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells: cells.map((row) => row.map((v) => (v == null ? null : -v))), fmt: (v) => (v == null ? '—' : (-v).toFixed(0) + 'y'), xfmt: (v) => '$' + Math.round(v / 1000) + 'k', yfmt: (v) => v.toFixed(0) + '%', xName: t('view.ytf.label.annual_expenses_usd') || 'Expenses', yName: t('view.ytf.label.annual_return_pct') || 'Return' });
}

File: crates/traderview-core/src/years_to_fi.rs
//! Years to financial independence — the savings-rate-to-timeline math.
//!
//! Financial independence is when your portfolio can fund your spending
//! indefinitely at a safe withdrawal rate: the **FI number** = annual
//! expenses / SWR (4% → 25× expenses). Starting from current savings and
//! adding the gap between income and expenses each year (grown at an
//! expected return), the years-to-FI is when the balance reaches that
//! number. The dominant lever is the **savings rate** — what fraction of
//! income you keep — far more than the return.
//!
//! Pure compute (year-by-year, end-of-year contributions, capped at 100y).

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct YearsToFiInput {
    pub current_savings_usd: f64,
    pub annual_income_usd: f64,
    pub annual_expenses_usd: f64,
    pub annual_return_pct: f64,
    /// Safe withdrawal rate defining the FI number (default 4%).
    pub safe_withdrawal_rate_pct: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct YearsToFiResult {
    /// Income − expenses: what's saved each year.
    pub annual_savings_usd: f64,
    pub savings_rate_pct: f64,
    /// Annual expenses / SWR — the portfolio needed to retire.
    pub fi_number_usd: f64,
    /// Years until the balance reaches the FI number; `None` if never (≤100y).
    pub years_to_fi: Option<u32>,
    /// True when current savings already meet the FI number.
    pub already_fi: bool,
}

const HORIZON: u32 = 100;

pub fn analyze(i: &YearsToFiInput) -> YearsToFiResult {
    let r = i.annual_return_pct / 100.0;
    let swr = if i.safe_withdrawal_rate_pct > 0.0 { i.safe_withdrawal_rate_pct } else { 4.0 };
    let annual_savings = i.annual_income_usd - i.annual_expenses_usd;
    let savings_rate = if i.annual_income_usd > 0.0 {
        annual_savings / i.annual_income_usd * 100.0
    } else {
        0.0
    };
    let fi_number = i.annual_expenses_usd / (swr / 100.0);

    let mut balance = i.current_savings_usd;
    let already_fi = balance >= fi_number;
    let mut years = if already_fi { Some(0u32) } else { None };

    if years.is_none() {
        for y in 1..=HORIZON {
            balance = balance * (1.0 + r) + annual_savings;
            if balance >= fi_number {
                years = Some(y);
                break;
            }
        }
    }

    YearsToFiResult {
        annual_savings_usd: annual_savings,
        savings_rate_pct: savings_rate,
        fi_number_usd: fi_number,
        years_to_fi: years,
        already_fi,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn inp(savings: f64, income: f64, expenses: f64, ret: f64) -> YearsToFiInput {
        YearsToFiInput {
            current_savings_usd: savings,
            annual_income_usd: income,
            annual_expenses_usd: expenses,
            annual_return_pct: ret,
            safe_withdrawal_rate_pct: 4.0,
        }
    }

    #[test]
    fn fi_number_is_expenses_over_swr() {
        // 40k / 4% = 1,000,000.
        let r = analyze(&inp(0.0, 80_000.0, 40_000.0, 0.0));
        assert!((r.fi_number_usd - 1_000_000.0).abs() < 1e-6);
    }

    #[test]
    fn savings_rate_and_annual_savings() {
        let r = analyze(&inp(0.0, 80_000.0, 40_000.0, 0.0));
        assert!((r.annual_savings_usd - 40_000.0).abs() < 1e-6);
        assert!((r.savings_rate_pct - 50.0).abs() < 1e-9);
    }

    #[test]
    fn already_fi_is_zero_years() {
        // 1M saved, FI number 1M → already there.
        let r = analyze(&inp(1_000_000.0, 80_000.0, 40_000.0, 5.0));
        assert!(r.already_fi);
        assert_eq!(r.years_to_fi, Some(0));
    }

    #[test]
    fn zero_return_is_linear() {
        // FI 750k, save 20k/yr, no return → 38 years (37×20k=740k<750k, 38×20k=760k).
        let r = analyze(&inp(0.0, 50_000.0, 30_000.0, 0.0));
        assert!((r.fi_number_usd - 750_000.0).abs() < 1e-6);
        assert_eq!(r.years_to_fi, Some(38));
    }

    #[test]
    fn higher_savings_rate_fewer_years() {
        let lean = analyze(&inp(0.0, 80_000.0, 60_000.0, 5.0)); // save 20k
        let aggressive = analyze(&inp(0.0, 80_000.0, 30_000.0, 5.0)); // save 50k
        assert!(aggressive.years_to_fi.unwrap() < lean.years_to_fi.unwrap());
    }

    #[test]
    fn higher_return_fewer_years() {
        let low = analyze(&inp(0.0, 80_000.0, 40_000.0, 3.0));
        let high = analyze(&inp(0.0, 80_000.0, 40_000.0, 8.0));
        assert!(high.years_to_fi.unwrap() < low.years_to_fi.unwrap());
    }

    #[test]
    fn expenses_at_or_above_income_never_via_saving() {
        // No savings, no return → never reaches FI within the horizon.
        let r = analyze(&inp(0.0, 50_000.0, 50_000.0, 0.0));
        assert!(r.annual_savings_usd.abs() < 1e-9);
        assert_eq!(r.years_to_fi, None);
    }

    #[test]
    fn return_compounds_toward_fi() {
        // With a return, FI arrives sooner than the linear no-return case.
        let flat = analyze(&inp(0.0, 80_000.0, 40_000.0, 0.0));
        let growth = analyze(&inp(0.0, 80_000.0, 40_000.0, 7.0));
        assert!(growth.years_to_fi.unwrap() < flat.years_to_fi.unwrap());
    }
}
